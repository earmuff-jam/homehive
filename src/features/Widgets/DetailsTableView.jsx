import { useEffect, useMemo, useState } from "react";
import {
  MaterialReactTable,
  useMaterialReactTable,
} from "material-react-table";
import { Stack } from "@mui/material";
import RowHeader from "src/common/RowHeader/RowHeader";
import { noramlizeDetailsTableData } from "src/features/Widgets/utils";
import dayjs from "dayjs";
import relativeTime from "dayjs/plugin/relativeTime";
import EmptyComponent from "common/EmptyComponent";

dayjs.extend(relativeTime);

const DetailsTableView = ({ label, caption }) => {
  const [tableData, setTableData] = useState([]);

  useEffect(() => {
    const draftData = JSON.parse(localStorage.getItem("pdfDetails"));
    const draftInvoiceStatus = JSON.parse(
      localStorage.getItem("invoiceStatus")
    );

    let formattedData;
    formattedData = { ...draftData };

    if (draftInvoiceStatus) {
      formattedData = { ...formattedData, invoice_status: draftInvoiceStatus };
    }

    if (draftData) {
      const data = noramlizeDetailsTableData([formattedData]);
      setTableData(data);
    }
  }, []);

  const columns = useMemo(
    () => [
      {
        accessorKey: "category",
        header: "Invoice Type",
        size: 200,
        Cell: ({ cell }) => (cell.getValue() ? cell.getValue() : "-"),
      },
      {
        accessorKey: "start_date",
        header: "Start Month",
        size: 150,
      },
      {
        accessorKey: "end_date",
        header: "End Month",
        size: 150,
      },
      {
        accessorKey: "total",
        header: "Total Collected",
        size: 150,
        Cell: ({ cell }) => `$${cell.getValue()}`,
      },
      {
        accessorKey: "invoice_status",
        header: "Invoice Status",
        size: 100,
        Cell: ({ cell }) => (cell.getValue() ? cell.getValue() : "-"),
      },
      {
        accessorKey: "payment_method",
        header: "Payment method",
        size: 150,
        Cell: ({ cell }) => (cell.getValue() ? cell.getValue() : "-"),
      },
      {
        accessorKey: "updated_on",
        header: "Updated on",
        size: 150,
        Cell: ({ cell }) => dayjs(cell.getValue()).fromNow(),
      },
    ],
    []
  );

  const table = useMaterialReactTable({
    columns,
    data: tableData,
    enableColumnActions: false,
    enableTopToolbar: false,
    initialState: {
      density: "comfortable",
    },
    renderEmptyRowsFallback: () => <EmptyComponent />,
    mrtTheme: (theme) => ({
      baseBackgroundColor: theme.palette.transparent.main,
    }),
    muiTableContainerProps: {
      sx: {
        maxHeight: "16rem",
        boxShadow: "none",
      },
    },
    muiTablePaperProps: {
      elevation: 0,
      sx: {
        boxShadow: "none",
      },
    },
  });

  return (
    <Stack spacing={2}>
      <RowHeader
        title={label}
        caption={caption}
        sxProps={{
          textAlign: "left",
          fontWeight: "bold",
          color: "text.secondary",
        }}
      />
      <MaterialReactTable table={table} />
    </Stack>
  );
};

export default DetailsTableView;
