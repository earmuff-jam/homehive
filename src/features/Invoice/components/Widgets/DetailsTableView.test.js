import React from "react";
import { render, screen } from "@testing-library/react";
import DetailsTableView from "./DetailsTableView";

// Mock common components and libraries
jest.mock("common/RowHeader/RowHeader", () => ({
  __esModule: true,
  default: ({ title, caption }) => (
    <div data-testid="row-header">
      <h1>{title}</h1>
      <p>{caption}</p>
    </div>
  ),
}));

jest.mock("common/EmptyComponent", () => ({
  __esModule: true,
  default: () => <div data-testid="empty">Empty</div>,
}));

jest.mock("features/Invoice/utils/utils", () => ({
  noramlizeDetailsTableData: jest.fn((data) => data),
}));

jest.mock("material-react-table", () => ({
  MaterialReactTable: ({ table }) => (
    <div data-testid="mock-table">Mock Table ({table.data.length} rows)</div>
  ),
  useMaterialReactTable: jest.fn((config) => config),
}));

describe("DetailsTableView component", () => {
  beforeEach(() => {
    localStorage.clear();
    jest.clearAllMocks();
  });

  it("renders correctly and matches snapshot", () => {
    const { asFragment } = render(
      <DetailsTableView label="Invoice Details" caption="Summary view" />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders RowHeader with label and caption", () => {
    render(<DetailsTableView label="Invoice Details" caption="Summary view" />);
    expect(screen.getByText("Invoice Details")).toBeInTheDocument();
    expect(screen.getByText("Summary view")).toBeInTheDocument();
  });

  it("renders MaterialReactTable mock", () => {
    render(<DetailsTableView label="Invoice Details" caption="Summary view" />);
    expect(screen.getByTestId("mock-table")).toBeInTheDocument();
  });
});
