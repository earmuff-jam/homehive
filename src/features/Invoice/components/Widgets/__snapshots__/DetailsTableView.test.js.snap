// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsTableView component renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root css-1sazv7p-MuiStack-root"
    data-tour="dashboard-7"
  >
    <div
      data-testid="row-header"
    >
      <h1>
        Invoice Details
      </h1>
      <p>
        Summary view
      </p>
    </div>
    <div
      data-testid="mock-table"
    >
      Mock Table (0 rows)
    </div>
  </div>
</DocumentFragment>
`;
